Add 'Trash in, Treasure out' blog post

hdoordt opened the pull request:
Companion repo: https://github.com/mainmatter/trash-in-treasure-out/

Todo:

- [x] add links to steps
- [x] Add short explanation of title (parody on 'garbage in, garbage out')
- [x] Do something about the mermaid diagram. Support for that on the site would be cool. @marcoow any ideas?

hdoordt commented:
I don't know what's going on with CI, but it doesn't seem like the failing jobs are caused by this addition?

BobrImperator commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,744 @@
+# Trash in, treasure out
```
The CI fails because the post doesn't contain frontmatter data at the top of the file.

See [example](https://github.com/mainmatter/mainmatter.com/blob/master/src/posts/2024-10-30-introducing-sheepdogjs.md?plain=1)

marcoow commented:
about Mermaid, you can ask @umbrellas – I guess there should be a plugin or so for 11ty

marcoow commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,744 @@
+data-carrying enums, closures, visibility specifiers, _explicit_ conversions, and
+much more. These are neat features that make performant, low-level programming 
+feel as ergonomic as high-level languages. Sure, Rust has a learning curve, and
+thatlearning curve is a result of Rust's tendency to make complexity really _in
```
```suggestion
that learning curve is a result of Rust's tendency to make complexity really _in
```

marcoow commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,744 @@
+
+Make no mistake, every piece of software is complex: it has to run on
+computers, which, especially nowadays are complex beasts. And writing
+software with highly optimized languages with manual memory management like
+C, C++, or Rust requires knowledge of all kinds of subtleties. Rust makes these
```
I find this confusing as it sounds like Rust had manual management as well?

marcoow replied:
I see what you mean but is that really how people talk about Rust's approach? I've always told people it does explicitly not require manual memory management like C or C++ but automatic compile time mm (and you need to follow certain rules to enable that).

Anyway, you've been in the Rust ecosystem for longer and much deeper than I have – if this is how people think/talk about this, I'm fine.

marcoow commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,744 @@
+  information.
+
+## Ticket to heaven
+We'll need a case to show how all this works, and since Mainmatter loves the
+travel industry, let's write up an API for booking train tickets.
```
```suggestion
We'll need a case to show how all this works, and since Mainmatter [loves the
travel industry](/travel/), let's write up an API for booking train tickets.
```

can't hurt to link to the landing page for context

marcoow commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,744 @@
+look!*
+
+
+[step 0]: todo
+[step 1]: todo
+[step 2]: todo
+[step 3]: todo
```
I guess these should be links to the example repo?

hdoordt replied:
Yup. But I might still do some rebasing. I expect to update these next friday

marcoow replied:
let's get this out before Friday

hdoordt replied:
In that case, someone else will have to do the links as I'm off until Friday. The commit history should on the companion repo should be clear enough. I guess it's fine if we update the links later it need be

marcoow replied:
Oh, I see – in that case I guess let's get this out on Friday

marcoow reviewed (commented):
This is a great post! Let's get it out asap. @IdeliaRisella can handle things like adding `og:image`, scheduling the release, etc.

If we can't figure out a solution for supporting Mermaid diagrams easily, let's just use an image instead for now.

IdeliaRisella commented:
@marcoow I have already sent the og:image to @hdoordt yesterday, let me know if you need anymore help. Meanwhile, I can help prepare the publication on social media on Friday.

hdoordt commented:
I replaced the mermaid source with a rendered SVG, and did some other fixes.

I added the metadata as suggested by @BobrImperator, but CI still seems to be failing

marcoow reviewed (commented):
@hdoordt you need to add your author profile image (named `hdoordt.jpg`) in https://github.com/mainmatter/mainmatter.com/tree/master/static/assets/images/authors

hdoordt commented:
@marcoow Most of the checks seem to pass right now. The 'Craw links' check seems to fail on some image that's not related to this PR not being there, and the others need manual interaction.

I looked at the preview and it looks good now to me

marcoow commented on src/posts/2024-12-02-trash-in-treasure-out.md:
```diff
@@ -0,0 +1,585 @@
+
+In the [introduction](#trash-in-treasure-out), I listed three pieces of advice:
+
+> 1. Encode the semantics/states of your application in the type system and your API.
+> 2. Ensure input gets parsed into rigid structs before acceptance.
+> 3. Ensure output gets encoded in the correct format and doesn’t leak (sensitive) information.
```
not sure this should look like this?

![Bildschirmfoto 2024-12-06 um 15 13 12](https://github.com/user-attachments/assets/708a0874-5f1b-43de-8a9f-9b0c835c1663)

hdoordt replied:
I guess? It's a quote after all. I don't particularly hate it
